packages/core/src/interaction-output.ts

```diff
@@ -79,43 +79,47 @@ export class InteractionOutput implements WoT.InteractionOutput {
 
     async value<T extends WoT.DataSchemaValue>(): Promise<T> {
         // the value has been already read?
-        if (this.parsedValue !== undefined) {
-            return this.parsedValue as T;
+        if (this.#value !== undefined) {
+            return this.#value as T;
         }
 
         if (this.dataUsed) {
-            throw new Error("Can't read the stream once it has been already used");
+            throw new NotReadableError("Can't read the stream once it has been already used");
+        }
+
+        if (this.form == null) {
+            throw new NotReadableError("No form defined");
+        }
+
+        if (this.schema == null || this.schema.type == null) {
+            throw new NotReadableError("No schema defined");
         }
 
         // is content type valid?
-        if (!this.form || !ContentSerdes.get().isSupported(this.content.type)) {
-            const message = !this.form ? "Missing form" : `Content type ${this.content.type} not supported`;
+        if (!ContentSerdes.get().isSupported(this.content.type)) {
+            const message = `Content type ${this.content.type} not supported`;
             throw new NotSupportedError(message);
         }
 
         // read fully the stream
-        const data = await this.content.toBuffer();
+        const bytes = await this.content.toBuffer();
         this.dataUsed = true;
-        this.buffer = data;
+        this.buffer = bytes;
 
         // call the contentToValue
-        // TODO: should be fixed contentToValue MUST define schema as nullable
-        const value = ContentSerdes.get().contentToValue({ type: this.content.type, body: data }, this.schema ?? {});
-
-        // any data (schema)?
-        if (this.schema) {
-            // validate the schema
-            const validate = ajv.compile<T>(this.schema);
-
-            if (!validate(value)) {
-                debug(`schema = ${util.inspect(this.schema, { depth: 10, colors: true })}`);
-                debug(`value: ${value}`);
-                debug(`Errror: ${validate.errors}`);
-                throw new DataSchemaError("Invalid value according to DataSchema", value as WoT.DataSchemaValue);
-            }
+        const json = ContentSerdes.get().contentToValue({ type: this.content.type, body: bytes }, this.schema);
+
+        // validate the schema
+        const validate = ajv.compile<T>(this.schema);
+
+        if (!validate(json)) {
+            debug(`schema = ${util.inspect(this.schema, { depth: 10, colors: true })}`);
+            debug(`value: ${json}`);
+            debug(`Errror: ${validate.errors}`);
+            throw new DataSchemaError("Invalid value according to DataSchema", json as WoT.DataSchemaValue);
         }
 
-        this.parsedValue = value;
-        return this.parsedValue as T;
+        this.#value = json;
+        return this.#value as T;
```

Comment:
Should we maybe include an additional check to make sure that the return value is really of type `T` here?

Reply:
the validate function does already all the checks, and the `json` variable is type of `T`. Maybe should we re-write it as:
```suggestion
        this.#value = json;
        return json;
```
?

Reply:
Incorporated the feedback in 1148fab.